File: 0238.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        M1 = [1 for x in range(len(nums) + 1)]
        M2 = [1 for x in range(len(nums) + 1)]
        nums2 = []
        
        for i in range(len(nums)):
            M1[i] = M1[i-1] * nums[i]
            
        for i in range(len(nums) - 1, -1, -1):
            M2[i] = M2[i+1] * nums[i]
                        
        for i in range(len(nums)):
            nums2.append(M1[i-1] * M2[i+1])
        
        return nums2

File: leetcode_75/array_string/0238.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var productExceptSelf = function (nums) {
  let prefixProduct = [];
  let suffixProduct = [];
  let ans = [];

  for (let i = 0; i < nums.length; i++) {
    let suffixIdx = nums.length - 1 - i;
    prefixProduct.push(
      prefixProduct.length
        ? nums[i] * prefixProduct[prefixProduct.length - 1]
        : nums[i]
    );
    suffixProduct.unshift(
      suffixProduct.length
        ? nums[suffixIdx] * suffixProduct[0]
        : nums[suffixIdx]
    );
  }

  for (let i = 0; i < nums.length; i++) {
    let productBefore = i > 0 ? prefixProduct[i - 1] : 1;
    let productAfter = i < nums.length - 1 ? suffixProduct[i + 1] : 1;
    ans.push(productBefore * productAfter);
  }

  return ans;
};

let nums = [1, 2, 3, 4];
let ans = productExceptSelf(nums);
console.log(ans);
